diff --git a/setup.py b/setup.py
index 7267529..098a6bb 100644
--- a/setup.py
+++ b/setup.py
@@ -37,7 +37,7 @@ requirements = [
     "idna>=2.0",
     "pyasn1>=0.1.8",
     "six>=1.4.1",
-    "setuptools>=11.3",
+    "setuptools",
 ]
 setup_requirements = []
 
diff --git a/src/cryptography/hazmat/backends/__init__.py b/src/cryptography/hazmat/backends/__init__.py
index 96a431d..256fee3 100644
--- a/src/cryptography/hazmat/backends/__init__.py
+++ b/src/cryptography/hazmat/backends/__init__.py
@@ -17,7 +17,12 @@ def _available_backends():
 
     if _available_backends_list is None:
         _available_backends_list = [
-            ep.resolve()
+            # setuptools 11.3 deprecated support for the require parameter to
+            # load(), and introduced the new resolve() method instead.
+            # This can be removed if/when we can assume setuptools>=11.3. At
+            # some point we may wish to add a warning, to push people along,
+            # but at present this would result in too many warnings.
+            ep.resolve() if hasattr(ep, "resolve") else ep.load(require=False)
             for ep in pkg_resources.iter_entry_points(
                 "cryptography.backends"
             )
